### PR TITLE
[CPU] Deconvolution int8 support

### PR DESCRIPTION
**Pull request by antonvor**
### Tickets:
 - *48331*
 
PR in oneDNN: openvinotoolkit/oneDNN#50
PR with tests: #5348
Builds: 624, 625

### TODO

- [ ] support post ops for int8 deconvolution (UPD: will be added in this PR https://github.com/openvinotoolkit/openvino/pull/5683)

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp [hunk 1]
```diff
@@ -196,36 +329,47 @@ void MKLDNNDeconvolutionNode::createDescriptor(const std::vector<InferenceEngine
...
-        descs_bwd.push_back(deconv_desc);
-
-        auto fwd_conv_pd = std::make_shared<convolution_forward::primitive_desc>(*conv_desc, getEngine(), true);
-        if (fwd_conv_pd->get(true) == nullptr)
-            continue;
-
-        descs.emplace_back(deconv_desc, fwd_conv_pd);
+    auto convert = [] (const std::vector<ptrdiff_t>& orig_dims) {
+        return memory::dims(orig_dims.begin(), orig_dims.end());
+    };
```
Comment (maxnick) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp:334]:
```suggestion
    auto convertDims = [] (const std::vector<ptrdiff_t>& orig_dims) {
        return memory::dims(orig_dims.begin(), orig_dims.end());
    };
```

File: inference-engine/src/mkldnn_plugin/mkldnn_descriptor.h [hunk 2]
```diff
@@ -15,6 +15,13 @@ class MKLDNNDescriptor {
 
     MKLDNNDescriptor(std::shared_ptr<mkldnn::convolution_backward_data::desc> desc,
                      std::shared_ptr<mkldnn::convolution_forward::primitive_desc> prim);
+
+    explicit MKLDNNDescriptor(std::shared_ptr<mkldnn::deconvolution_forward::desc> desc);
+    operator std::shared_ptr<mkldnn::deconvolution_forward::desc>();
+
+    MKLDNNDescriptor(std::shared_ptr<mkldnn::deconvolution_forward::desc> desc,
+            std::shared_ptr<mkldnn::deconvolution_forward::primitive_desc> prim);
```
Comment (maxnick) [inline inference-engine/src/mkldnn_plugin/mkldnn_descriptor.h:23]: Are we sure we need this constructor?

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp [hunk 3]
```diff
@@ -79,23 +85,119 @@ MKLDNNDeconvolutionNode::MKLDNNDeconvolutionNode(const std::shared_ptr<ngraph::N
...
+    }
+    for (int i = 2 + withGroups; i < dimsForBlockedDesc.size(); i++)
+        orderForBlockedDesc.push_back(i);
+
+    BlockingDesc blkDesc(dimsForBlockedDesc, orderForBlockedDesc);
+    InferenceEngine::TensorDesc tensorDesc(blb->getTensorDesc().getPrecision(), dims, blkDesc);
+
+    auto fillInternalBlob = [&](char *data, size_t intBuffSize) {
+        size_t offset = blb->byteSize();
+        checkSize(intBuffSize, offset);
```
Comment (maxnick) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp:136]: the `checkSize` lambda is used once. Suggest inline the code.

**Review by maxnick — CHANGES_REQUESTED** [review]

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp [hunk 4]
```diff
@@ -79,23 +85,119 @@ MKLDNNDeconvolutionNode::MKLDNNDeconvolutionNode(const std::shared_ptr<ngraph::N
...
+        cpu_memcpy_s(data, intBuffSize, blb->cbuffer(), blb->byteSize());
+    };
+
+    Blob::Ptr internalBlob = InferenceEngine::make_shared_blob<int8_t>(tensorDesc);
+    internalBlob->allocate();
+    char *data = internalBlob->buffer();
```
[the remark this answers is not in the document]
Comment (antonvor) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp:130]: added

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp [hunk 5]
```diff
@@ -79,23 +85,119 @@ MKLDNNDeconvolutionNode::MKLDNNDeconvolutionNode(const std::shared_ptr<ngraph::N
...
+    size_t intBuffSize = internalBlob->byteSize();
+
+    fillInternalBlob(data, intBuffSize);
```
[the remark this answers is not in the document]
Comment (antonvor) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp]: thanks, fixed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp [hunk 6]
```diff
@@ -79,23 +85,119 @@ MKLDNNDeconvolutionNode::MKLDNNDeconvolutionNode(const std::shared_ptr<ngraph::N
...
-    auto outputDataType = MKLDNNExtensionUtils::IEPrecisionToDataType(precision);
+    isInt8 = canBeExecutedInInt8();
+
+    InferenceEngine::Precision inPrecision = getOriginalInputPrecisionAtPort(0);
+    InferenceEngine::Precision outPrecision = getOriginalOutputPrecisionAtPort(0);
+    if (!isInt8) {
+        if (!one_of(inPrecision, InferenceEngine::Precision::FP32, InferenceEngine::Precision::BF16))
+            inPrecision = InferenceEngine::Precision::FP32;
+        if (!one_of(outPrecision, InferenceEngine::Precision::FP32, InferenceEngine::Precision::BF16))
+            outPrecision = InferenceEngine::Precision::FP32;
```
[the remark this answers is not in the document]
Comment (antonvor) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp:191]: Yes you are right. We must return to this issue in the future. But not in this PR Thanks for your comment!.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp (hunk 1 above)
Comment (antonvor) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp]: fixed, thanks

File: inference-engine/src/mkldnn_plugin/mkldnn_descriptor.h (hunk 2 above)
Comment (antonvor) [reply inference-engine/src/mkldnn_plugin/mkldnn_descriptor.h]: removed, thanks

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp (hunk 3 above)
Comment (antonvor) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp]: fixed, thanks

**Comment by antonvor** [comment]
@maxnick thanks for your review!
@dmitry-gorokhov take a look please

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (maxnick) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp:191]: Maybe it makes sense to create a ticket?

**Review by maxnick — APPROVED** [review]

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (dmitry-gorokhov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_deconv_node.cpp:191]: Should be aligned with BF16+INT8 activity

**Review by dmitry-gorokhov — APPROVED** [review]
Comments to be resolved in separate PR

**Pull request merged**